### PR TITLE
[FLINK-30863][state] Register local recovery files of changelog before notifyCheckpointComplete()

### PR DESCRIPTION
**Pull request by fredia**
## What is the purpose of the change

If TM is materialized before receiving confirm(), the previously uploaded queue in `FsStateChangelogWriter` will be cleared, so the local files of the completed checkpoint will not be registered again, while the JM owned files are registered before confirm(), and do not depend on the uploaded queue, so the local files are deleted, and the DFS files are still there. 
This PR makes the local recovery files registered before confirm() to avoid this problem. 

## Brief change log
  - Replace registering local files in confirm() with  registering local files in buildSnapshotResult().
 

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change is already covered by existing tests, such as `ChangelogLocalRecoveryITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by fredia** [comment]
Hi @rkhachatryan, could you please help take a look?

**Comment by jiexray** [comment]
Hello @fredia, thank you for creating this PR. I have some concern about the removal of `LocalChangelogRegistryImpl#prune()`, please have a look.

In my opinion, the main purpose of `LocalChangelogRegistryImpl#prune()` is to proactively remove the useless state uploaded by an aborted checkpoint. More precisely, the useless state is uploaded by an aborted checkpoint and is not shared by previous checkpoints. However, the current removal mechanism may delete the state of a previous checkpoint. I think the root cause of the undesirable removal is cause by the following code snippet:
```
public void register(StreamStateHandle handle, long checkpointID) {
    handleToLastUsedCheckpointID.compute(
            handle.getStreamStateHandleID(),
            (k, v) -> {
                if (v == null) {
                    return Tuple2.of(handle, checkpointID);
                } else {
                    Preconditions.checkState(handle.equals(v.f0));
                    // !Here: already existed `handle` will be removed by an aborted checkpoint with `checkpointID`
                    return Tuple2.of(handle, Math.max(v.f1, checkpointID));
                }
            });
}
```
This PR finds a way to fix the aforementioned root cause (`Math.max(v.f1, checkpointID)`) by removing the `LocalChangelogRegistryImpl#prune()`.

Therefore, the useless state uploaded by an aborted checkpoint (denoted by CHK) will be passively removed by `LocalChangelogRegistryImpl#discardUpToCheckpoint(CHK + 1)`. I was wondering if the useless state could be the blocker of the next invocation of `LocalChangelogRegistryImpl#discardUpToCheckpoint(CHK + 1)`. For example, maybe, the useless state takes up some disk space and leads to other failure. 

By the way, I am thinking whether we could change the registering machanism in `LocalChangelogRegistryImpl#register` to get around the undesirable removal. The objective of this design is to stick on proactively removing the useless state upon checkpoint aborting. For example, support that `CHK` is the current checkpoint. 
 - Method-1: We temporarily store the `(CHK, StreamStateHandle)` in a new map. When registering `CHK+1`, we merge the mapping of `(CHK, StreamStateHandle)` to `handleToLastUsedCheckpointID`. 
 - Method-2: Use reference counting for `handleToLastUsedCheckpointID`.

File: flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java
Comment (fredia) [reply flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java]: The registration in the `confirm()` was moved to ChangelogKeyedStatebacked now, during one checkpoint, one local handle is only registered once.

This change would affect `LocalChangelogRegistryImpl#prune()`, before this changes, the `prune()` wouldn't cause `fileNotFound`, because `notifyAbort()` is impossible to come after `notifyComplete()` on TM(The lastUsedCheckpointId of previous chk only update on `notifyComplete()`).  So I plan to fix the problem of `abort()` in another commit.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:204]: I think it is better to let the test code adapt to the interface change.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:213]: we can add checkpoint to debug log

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:391]: since already register local handle in persist method , this line can be deleted.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:55]: Control transfer is no longer when ChangelogKeyedStateBackend#buildSnapshotResult

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:372]: In order to delete useless local files as soon as possible, discardUpToCheckpoint should not be removed from here.

**Review by zoltar9264 — CHANGES_REQUESTED** [review]
Thanks for your work @fredia , I leave some comments , PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java:64]: No, normally there is no such case: new handle doesn't equal v.getHandle.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java:91]: Agreed, the pre-emptive upload files would not be re-upload, so those aborted files may be referenced by subsequent checkpoints.  But in case of many subsequent aborted checkpoints, local state accumulates, all aborted local state will not be deleted until the next completed materialization. 

Considering that the priority of correctness is higher than that of space amplification, I first deleted this method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistry.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistry.java:63]: Previously, `discardUpToCheckpoint` was called in both `confirm()` and `subsume()`, and in fact this is a duplicate, now it is only called in `subsume()`, and the logic in `discardUpToCheckpoint` also changed (`<` -> `<=`).
If `state.checkpoints.num-retained >1`, this will retain more local files that are not needed, because local recovery only needs one previous chk. 

I fixed this by remove `subsume` and only call `discardUpToCheckpoint` when `confirm()`.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogStorage.java:170]: Fixed,replace with `IOUtils.closeQuietly` .

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:204]: The tests related to the Local file have been modified accordingly, for other tests I left them as they are.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:391]: Fixed.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:55]: Fixed.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:372]: Fixed, retain the call of `discardUpToCheckpoint` here and delete the call in `subsume()`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java
Comment (fredia) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/changelog/LocalChangelogRegistryImpl.java]: Since we don't need  `LocalChangelogRegistry#prune(long)`, I deleted the inner class.

**Comment by fredia** [comment]
@zoltar9264 Thanks for the review and suggestion, I updated the PR based on the comment. Currently, the `LocalChangelogRegistryImpl#prune()` is removed.

**Review by zoltar9264 — COMMENTED** [review]
Thanks for update the pr @fredia , most fine for me, I left some minor suggestions, PLTA.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:204]: I think all of that should be modified. Since that all test code can easily adapt to the new interface, I don't see the need to keep this method for test code alone. In my opinion this breaks the legibility of the code.

File: flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java:366]: Deleted it, it was used to highlight `notifySubsume()`.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:308]: You may have misunderstood, I mean it only needs to be registered when result.localStreamHandle is not null.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (zoltar9264) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:308]:
```suggestion
                                if (result.localStreamHandle != null) {
                                    localChangelogRegistry.register(
                                            result.localStreamHandle, checkpointId);
                                }
```

**Review by zoltar9264 — COMMENTED** [review]
Thanks for update the pr, there is a misunderstand of my previous comment, PTAL.

**Comment by fredia** [comment]
@zoltar9264 Thanks for the review, I refactored all `persist(sqn)` to `persist(sqn,long)`, and sorry for the misunderstood, I added null check before registering local handle as you suggested.

**Comment by zoltar9264** [comment]
LGTM @fredia , thanks for your work.

**Comment by zoltar9264** [comment]
Hi @curcur and @rkhachatryan , can you help us review this pr ?

**Comment by zoltar9264** [comment]
Hi @Myasuka , may I request you help us review this pr ?

**Comment by Myasuka** [comment]
> Hi @Myasuka , may I request you help us review this pr ?

Sure, I'll take a review today.

File: flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java
Comment (Myasuka) [inline flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java:252]: I tried to revert the code changes in this PR, this unit test could still pass. It seems the added test is useless.

**Review by Myasuka — COMMENTED** [review]
Please take a look on my comments.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (zoltar9264) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: Yes, the cleaning up of local files depends on the newly completed checkpoint, local files will accumulate if no checkpoint could complete. IIUC, remote dstl file has same problem. I did encounter this problem in our environment. 
image
I suggest open a new ticket to solve this problem both in remote and local, WDYT @Myasuka ?

Comment (Myasuka) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: I think the problem does not exist for local checkpoints before this PR. If so, I am against to introduce anther new problem here.

Comment (zoltar9264) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: The problem of not cleaning up useless local files after truncate was not actively addressed before this PR. Previously, it was only aggressive to delete all files referenced by checkpoints that aborted.

File: flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java:252]: Thanks for the comments, as @zoltar9264 mentioned, this UT is used to verify that the newly changed code can correctly delete the local file. It doesn't have a bad effect on current code, I tend to keep it here.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: The current implementation can partially alleviate the issue of local file accumulation:
1. In the case that TM can exit normally(`ChangelogTaskLocalStateStore#dispose()` can be executed), all local files can be cleaned up. 
2. If `emptyDir` is used as the local disk, all files will be deleted when the pod exits.

So I think the problem with local dstl files is milder than remote dstl files. 

[Pruning ](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskStateManagerImpl.java#L210) other dstl local files during restore may alleviate this issue, WDYT?

File: flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java
Comment (Myasuka) [reply flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java:252]: If so, why we cannot introduce a UT would fail with code before this PR?

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (Myasuka) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: @zoltar9264 My concern is that we already know that this solution is not a good choice, we might need to refactor this in the near feature. Why we have to do the work twice?

On the other hand, since local disk has a much smaller space than DFS. Leaving files on the disk could make no space left problem more easily, which actually make the status worse.

File: flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/test/java/org/apache/flink/changelog/fs/FsStateChangelogWriterTest.java:252]: @Myasuka 
`testLocalFileAfterMaterialize()` would fail with code before this PR,  because the previous test of the local dstl files are not enough, `testLocalFileDiscard()` is a supplement.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java
Comment (zoltar9264) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: Thanks for explanation @Myasuka  , I agree that one shouldn't introduce a solution that will be refactored anytime soon. 
But I just think the approach in that PR is one that will still be appropriate in the long run. We can divide the deletion of dstl files into two categories :

1. delete by subsumed checkpoint 
        remote files : delete by jobmanager , nothing with taskmanager
        local files : delete by _LocalChangelogRegistry#discardUpToCheckpoint_ (this PR does exactly that)
2. delete by not used StateChangeSet ( also contains thats only used by aborted checkpoints ) 
        remote files and local files are in the same processing logic, both are processed by _TaskChangelogRegistry_ . Perhaps we can introduce a "StateChangeUsageTracker" component to find out those not used StateChangeSets based on checkpoint completion and materialization execution, and then notify TaskChangelogRegistry to release the reference count of the corresponding file.

In the method described above, the LocalChangelogRegistry is only responsible for the deletion of local files at subsumed checkpoint, as implemented in this PR. So I suggest approve this pr and open a new ticket to address deletion of not used both remote and local files .

Comment (Myasuka) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/DuplicatingStateChangeFsUploader.java:57]: I think a better way is that we should move to the Flink ticket and have a clean discussion on this problem with consideration of the whole plan, instead of just approving a half-baked PR.
What's more, we don't have a release deadline recently, and shall not be hurry and in this case.

**Review by zoltar9264 — COMMENTED** [review]
Thanks @fredia for updating the pr, most look good to me except a minor java doc description, PTAL.

**Comment by fredia** [comment]
@zoltar9264 Thanks for the detailed review, I addressed the comment and squashed some commits.

**Comment by zoltar9264** [comment]
LGTM, @Myasuka Do you have any other comments?

**Comment by fredia** [comment]
CI failed for https://issues.apache.org/jira/projects/FLINK/issues/FLINK-31962, rebased.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (Myasuka) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:404]: I don't think the `#discardUpToCheckpoint` has the same meaning as the previous `#prune`. 
`#prune` would only delete files with the specific checkpoint id, while `discardUpToCheckpoint` would delete files with checkpoint id smaller than the given one. And the `notifyCheckpointAbort` should tell the TM only abort files with that checkpoint id.

**Review by Myasuka — COMMENTED** [review]
I think the refactoring in the 3rd commit is different from previous logic. Please take a look at my comments.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java
Comment (fredia) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java:404]: As we discuss offline, because changelog state backend is incremental, when a checkpoint is aborted, all subsequent checkpoints that depend on this checkpoint are incomplete, It is equivalent to all previous checkpoints that are no longer useful. Therefore, we broaden the responsibility of `#abort`/`#prune` and delete all the previous useless local checkpoints.

**Review by Myasuka — APPROVED** [review]
LGTM.

**Comment by zoltar9264** [comment]
Thanks @Myasuka , can you help us merge this pr ? or we need more committer approval ?

**Comment by Myasuka** [comment]
> Thanks @Myasuka , can you help us merge this pr ? or we need more committer approval ?

Since I am not focusing on changelog state-backend currently, I think it's better to have another review from committers who run changelog state-backend in the production environment. Maybe @curcur or @rkhachatryan, and I think @StefanRRichter could also have some insights if possible :)

**Comment by zoltar9264** [comment]
Hi @rkhachatryan , @StefanRRichter , @curcur , can you help review the pr :)

**Pull request merged**
